File: native/android/app/src/main/kotlin/art/lazying/auspice/Engines.kt
package art.lazying.auspice

import android.content.Context
import androidx.concurrent.futures.await
import androidx.javascriptengine.JavaScriptIsolate
import androidx.javascriptengine.JavaScriptSandbox
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * The divination rules, called from Kotlin.
 *
 * Every rule in this product is written once, in TypeScript, under
 * `src/engines/`, with golden-file tests. `lazyoracle-engines.js` is that same
 * code built into one bundle with no DOM and no network, and this object is the
 * only thing between it and the native screens.
 *
 * It runs in `androidx.javascriptengine`, a JavaScript sandbox in its own
 * process with no document, no storage and no network — the counterpart of
 * JavaScriptCore on iOS. Writing the rules a second time in Kotlin would look
 * like less work for a week and then quietly disagree with the other two apps
 * about someone's day master, with no test able to catch it.
 */
object Engines {
    const val EXPECTED_VERSION = 1

    class Failure(message: String) : Exception(message)

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val lock = Mutex()
    private var isolate: JavaScriptIsolate? = null
    private var sandbox: JavaScriptSandbox? = null

    var startupError: String? = null
        private set

    val ready: Boolean get() = isolate != null

    /** Loads the bundle once, at launch. */
    suspend fun start(context: Context) = withContext(Dispatchers.IO) {
        lock.withLock {
            if (isolate != null) return@withLock
            try {
                if (!JavaScriptSandbox.isSupported()) {
                    throw Failure("This device's system WebView is too old to run the rules.")
                }
                val box = JavaScriptSandbox.createConnectedInstanceAsync(context.applicationContext).await()
                val iso = box.createIsolate()
                val source = context.assets.open("lazyoracle-engines.js").bufferedReader().use { it.readText() }

                if (box.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_EVALUATE_WITHOUT_TRANSACTION_LIMIT)) {
                    iso.evaluateJavaScriptAsync(source).await()
                } else if (box.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_PROVIDE_CONSUME_ARRAY_BUFFER) &&
                    box.isFeatureSupported(JavaScriptSandbox.JS_FEATURE_PROMISE_RETURN)
                ) {
                    // The bundle is larger than a binder transaction, so it is
                    // handed over as named data and read back inside the sandbox.
                    iso.provideNamedData("rules", source.toByteArray())
                    iso.evaluateJavaScriptAsync(
                        """
                        android.consumeNamedDataAsArrayBuffer("rules").then((buffer) => {
                          (0, eval)(new TextDecoder().decode(buffer));
                          return typeof LazyOracle === "object" ? "ok" : "missing";
                        })
                        """.trimIndent()
                    ).await()
                } else {
                    throw Failure("This device's system WebView cannot load a bundle this size.")
                }

                sandbox = box
                isolate = iso
                startupError = null
            } catch (error: Throwable) {
                startupError = error.message ?: error.toString()
            }
        }
    }

    /**
     * Runs one engine and hands back its `data`. A failure comes back as an
     * exception carrying the engine's own words, never as a silent empty result.
     */
    suspend fun evaluate(engine: String, input: JsonObject = JsonObject(emptyMap())): JsonElement =
        withContext(Dispatchers.IO) {
            val iso = isolate ?: throw Failure(startupError ?: "The rules are not loaded yet.")
            val request = json.encodeToString(
                JsonObject.serializer(),
                buildJsonObject {
                    put("engine", JsonPrimitive(engine))
                    put("input", input)
                }
            )
            val literal = json.encodeToString(JsonPrimitive.serializer(), JsonPrimitive(request))
            val returned = lock.withLock {
                iso.evaluateJavaScriptAsync("LazyOracle.evaluateJson($literal)").await()
            }
            val envelope = json.parseToJsonElement(returned).jsonObject
            if (envelope["ok"]?.jsonPrimitive?.content != "true") {
                throw Failure(envelope["error"]?.jsonPrimitive?.content ?: "$engine failed")
            }
            val version = envelope["version"]?.jsonPrimitive?.content?.toIntOrNull()
            if (version != null && version != EXPECTED_VERSION) {
                throw Failure("The rules bundle speaks version $version; this app expects $EXPECTED_VERSION.")
            }
            envelope["data"] ?: throw Failure("$engine returned nothing")
        }

    /** The same call, decoded into a Kotlin type. */
    suspend inline fun <reified T> evaluateAs(engine: String, input: JsonObject = JsonObject(emptyMap())): T {
        val data = evaluate(engine, input)
        return lenient.decodeFromJsonElement(data)
    }

    val lenient = Json { ignoreUnknownKeys = true; isLenient = true; coerceInputValues = true }
}

File: native/ios/Auspice/Engines.swift
import Foundation
import JavaScriptCore

/// The divination rules, called from Swift.
///
/// Every rule in this product — the four pillars, the almanac's tables, the
/// tarot draw, the hexagram cast, the natal chart — is written once, in
/// TypeScript, under `src/engines/`, with golden-file tests. `LazyOracle.js` is
/// that same code built into one bundle with no DOM and no network, and this
/// type is the only thing between it and the native screens.
///
/// A second implementation in Swift would look like less work for a week and
/// then quietly disagree with the web app about someone's day master, with no
/// test able to catch it. So: native interface, shared rules.
final class Engines {
    static let shared = Engines()

    enum Failure: LocalizedError {
        case bundleMissing
        case engineFailed(String)
        case badResult

        var errorDescription: String? {
            switch self {
            case .bundleMissing: "The rules bundle is missing from the app."
            case .engineFailed(let message): message
            case .badResult: "The rules returned something unreadable."
            }
        }
    }

    /// The contract version `engine-bridge.ts` declares. A mismatch means the
    /// bundle and this app were built from different commits.
    static let expectedVersion = 1

    private let queue = DispatchQueue(label: "art.lazying.auspice.engines", qos: .userInitiated)
    private var context: JSContext?
    private var loadFailure: Error?

    private init() {
        queue.sync {
            do {
                context = try Self.makeContext()
            } catch {
                loadFailure = error
            }
        }
    }

    private static func makeContext() throws -> JSContext {
        guard let url = Bundle.main.url(forResource: "lazyoracle-engines", withExtension: "js"),
              let source = try? String(contentsOf: url, encoding: .utf8) else {
            throw Failure.bundleMissing
        }
        guard let context = JSContext() else { throw Failure.bundleMissing }
        var thrown: String?
        context.exceptionHandler = { _, exception in
            thrown = exception?.toString() ?? "unknown JavaScript error"
        }
        context.evaluateScript(source, withSourceURL: url)
        if let thrown { throw Failure.engineFailed(thrown) }
        guard context.objectForKeyedSubscript("LazyOracle")?.isObject == true else {
            throw Failure.bundleMissing
        }
        return context
    }

    /// Runs one engine and hands back its `data`, already decoded.
    ///
    /// Engines are pure and fast — a full BaZi chart is well under a
    /// millisecond — so this is synchronous on a serial queue rather than
    /// async: making every screen await a chart would buy nothing and cost a
    /// frame of latency on every tap.
    func evaluate<T: Decodable>(_ engine: String, _ input: [String: Any] = [:], as type: T.Type) throws -> T {
        let data = try evaluate(engine, input)
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw Failure.engineFailed("\(engine): \(error.localizedDescription)")
        }
    }

    /// The same call, left as JSON, for the chat agent to hand to the model.
    func evaluate(_ engine: String, _ input: [String: Any] = [:]) throws -> Data {
        try queue.sync {
            if let loadFailure { throw loadFailure }
            guard let context, let bridge = context.objectForKeyedSubscript("LazyOracle") else {
                throw Failure.bundleMissing
            }
            let request = try JSONSerialization.data(
                withJSONObject: ["engine": engine, "input": input],
                options: []
            )
            guard let requestText = String(data: request, encoding: .utf8) else { throw Failure.badResult }

            var thrown: String?
            context.exceptionHandler = { _, exception in
                thrown = exception?.toString() ?? "unknown JavaScript error"
            }
            let returned = bridge.invokeMethod("evaluateJson", withArguments: [requestText])
            if let thrown { throw Failure.engineFailed("\(engine): \(thrown)") }
            guard let text = returned?.toString(), let payload = text.data(using: .utf8),
                  let envelope = try? JSONSerialization.jsonObject(with: payload) as? [String: Any] else {
                throw Failure.badResult
            }
            guard envelope["ok"] as? Bool == true else {
                throw Failure.engineFailed(envelope["error"] as? String ?? "\(engine) failed")
            }
            if let version = envelope["version"] as? Int, version != Self.expectedVersion {
                throw Failure.engineFailed("rules bundle speaks version \(version), this app expects \(Self.expectedVersion)")
            }
            guard let body = envelope["data"] else { throw Failure.badResult }
            // Some engines return a scalar: fengshui.sector returns "N", etc.
            // Without fragmentsAllowed, Foundation raises an Objective-C
            // exception here (not a Swift Error that try?/catch can handle).
            return try JSONSerialization.data(withJSONObject: body, options: [.fragmentsAllowed])
        }
    }

    /// Which engines the bundle carries, straight from `engines()`. Used by
    /// the chat agent to advertise its tools, and by the diagnostic line in
    /// Settings when a reading will not compute.
    func available() -> [String] {
        queue.sync {
            guard let bridge = context?.objectForKeyedSubscript("LazyOracle"),
                  let listed = bridge.invokeMethod("engines", withArguments: [])?.toArray() as? [String] else {
                return []
            }
            return listed
        }
    }
}
